File: graphql-endpoint/src/main/java/nl/openweb/graphql_endpoint/service/MoneyTransferService.java
package nl.openweb.graphql_endpoint.service;

import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import nl.openweb.data.ConfirmMoneyTransfer;
import nl.openweb.data.MoneyTransferConfirmed;
import nl.openweb.data.MoneyTransferFailed;
import nl.openweb.graphql_endpoint.kafka.CommandPublisher;
import nl.openweb.graphql_endpoint.mapper.MoneyTransferResultMapper;
import nl.openweb.graphql_endpoint.mapper.UuidMapper;
import nl.openweb.graphql_endpoint.model.MoneyTransferResult;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.reactivestreams.Publisher;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Component;
import reactor.core.publisher.EmitterProcessor;
import reactor.core.publisher.Flux;
import reactor.util.function.Tuple2;
import reactor.util.function.Tuples;

@Slf4j
@Component
@AllArgsConstructor
public class MoneyTransferService {

    private final CommandPublisher publisher;
    private final EmitterProcessor<Tuple2<String, MoneyTransferResult>> emitterProcessor = EmitterProcessor.create(false);
    private final Flux<Tuple2<String, MoneyTransferResult>> flux = Flux.from(emitterProcessor).share();

    @KafkaListener(id = "graphql-endpoint-money-transfer-consumer", topics = "${kafka.mtf-topic}")
    public void onMoneyTransfer(ConsumerRecord<String, Object> record) {
        log.info("received money transfer feedback from partition: {} at offset: {}",
                 record.partition(), record.offset());
        handleRecord(record);
    }


    private void handleRecord(ConsumerRecord<String, Object> record) {
        if (record.value() instanceof MoneyTransferConfirmed) {
            handleMoneyTransferConfirmed((MoneyTransferConfirmed) record.value());
        } else if (record.value() instanceof MoneyTransferFailed) {
            handleMoneyTransferFailed((MoneyTransferFailed) record.value());
        } else {
            log.warn(
                    "Unexpected item {} with class {}, expected either AccountCreationConfirmed or " +
                            "AccountCreationFailed",
                    record.value(), record.value().getClass().getName());
        }
    }

    private void handleMoneyTransferConfirmed(MoneyTransferConfirmed confirmed) {
        MoneyTransferResult result = MoneyTransferResultMapper.fromMoneyTransferConfirmed(confirmed);
        var item = Tuples.of(UuidMapper.fromAvroUuid(confirmed.getId()).toString(), result);
        emitterProcessor.onNext(item);
    }

    private void handleMoneyTransferFailed(MoneyTransferFailed failed) {
        MoneyTransferResult result = MoneyTransferResultMapper.fromMoneyTransferFailed(failed);
        var item = Tuples.of(UuidMapper.fromAvroUuid(failed.getId()).toString(), result);
        emitterProcessor.onNext(item);
    }

    public Publisher<MoneyTransferResult> transfer(long amount, String descr, String from, String to, String
            token,
            String username, String uuid) {
        ConfirmMoneyTransfer transfer = new ConfirmMoneyTransfer();
        transfer.setId(UuidMapper.fromString(uuid));
        transfer.setAmount(amount);
        transfer.setDescription(descr);
        transfer.setFrom(from);
        transfer.setTo(to);
        transfer.setToken(token);
        publisher.publish(username, transfer);
        return Flux.from(flux)
                .filter(x -> x.getT1().equals(uuid))
                .map(Tuple2::getT2);
    }
}

File: graphql-endpoint/src/main/kotlin/nl/openweb/graphql_endpoint/service/MoneyTransferService.kt
package nl.openweb.graphql_endpoint.service

import arrow.core.Tuple2
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.channels.BroadcastChannel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.consumeAsFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.map
import nl.openweb.data.ConfirmMoneyTransfer
import nl.openweb.data.MoneyTransferConfirmed
import nl.openweb.data.MoneyTransferFailed
import nl.openweb.graphql_endpoint.kafka.CommandPublisher
import nl.openweb.graphql_endpoint.logger
import nl.openweb.graphql_endpoint.model.MoneyTransferResult
import nl.openweb.graphql_endpoint.model.toMoneyTransferResult
import nl.openweb.graphql_endpoint.toAvroUuid
import nl.openweb.graphql_endpoint.toJavaUUID
import org.apache.kafka.clients.consumer.ConsumerRecord
import javax.inject.Singleton

@FlowPreview
@ExperimentalCoroutinesApi
@Singleton
class MoneyTransferService(private val publisher: CommandPublisher) {
    private val broadcast = BroadcastChannel<Tuple2<String, MoneyTransferResult>>(capacity = Channel.CONFLATED)
    private val log by logger()

    suspend fun handleRecord(record: ConsumerRecord<String, Any>) {
        if (record.value() is MoneyTransferConfirmed) {
            handleMoneyTransferConfirmed(record.value() as MoneyTransferConfirmed)
        } else if (record.value() is MoneyTransferFailed) {
            handleMoneyTransferFailed(record.value() as MoneyTransferFailed)
        } else {
            log.warn(
                    "Unexpected item {} with class {}, expected either AccountCreationConfirmed or " +
                            "AccountCreationFailed",
                    record.value(), record.value().javaClass.name)
        }
    }

    private suspend fun handleMoneyTransferConfirmed(confirmed: MoneyTransferConfirmed) {
        val result = confirmed.toMoneyTransferResult()
        val item = Tuple2(confirmed.id.toJavaUUID().toString(), result)
        broadcast.send(item)
    }

    private suspend fun handleMoneyTransferFailed(failed: MoneyTransferFailed) {
        val result = failed.toMoneyTransferResult()
        val item = Tuple2(failed.id.toJavaUUID().toString(), result)
        broadcast.send(item)
    }

    fun transfer(
        amount: Int,
        descr: String,
        from: String,
        to: String,
        token: String,
        username: String,
        uuid: String
    ): Flow<MoneyTransferResult> {
        val transfer = ConfirmMoneyTransfer()
        transfer.id = uuid.toAvroUuid()
        transfer.amount = amount.toLong()
        transfer.description = descr
        transfer.from = from
        transfer.to = to
        transfer.token = token
        publisher.publish(username, transfer)
        return broadcast
                .openSubscription()
                .consumeAsFlow()
                .filter { x: Tuple2<String, MoneyTransferResult> -> x.a == uuid }
                .map { obj: Tuple2<String, MoneyTransferResult> -> obj.b }
    }
}